src/aiida_quantumespresso/workflows/pw/base.py

```diff
@@ -405,21 +440,35 @@ def handle_known_unrecoverable_failure(self, calculation):
         exit_codes=[
             PwCalculation.exit_codes.ERROR_COMPUTING_CHOLESKY,
             PwCalculation.exit_codes.ERROR_DIAGONALIZATION_TOO_MANY_BANDS_NOT_CONVERGED,
+            PwCalculation.exit_codes.ERROR_S_MATRIX_NOT_POSITIVE_DEFINITE,
+            PwCalculation.exit_codes.ERROR_ZHEGVD_FAILED,
+            PwCalculation.exit_codes.ERROR_QR_FAILED,
+            PwCalculation.exit_codes.ERROR_EIGENVECTOR_CONVERGENCE,
+            PwCalculation.exit_codes.ERROR_BROYDEN_FACTORIZATION,
         ]
     )
     def handle_diagonalization_errors(self, calculation):
         """Handle known issues related to the diagonalization.
 
-        Switch to ``diagonalization = 'cg'`` if not already running with this setting, and restart from the charge
-        density. In case the run already used conjugate gradient diagonalization, abort.
+        We use the following strategy. When a diagonalization algorithm fails, we try using an other one
+        still not used. Conjugate gradient (CG) is kept as last option, as it is the slowest among the
+        available ones, but on the contrary it is the most stable as well, thus kept as `last resort`.
+
+        If also CG is failing, abort.
         """
-        if self.ctx.inputs.parameters['ELECTRONS'].get('diagonalization', None) == 'cg':
+        current_diagonalization = self.get_current_diagonalization()
+        if current_diagonalization == DiagonalizationType.CG:
             action = 'found diagonalization issues but already running with conjugate gradient algorithm, aborting...'
             self.report_error_handled(calculation, action)
             return ProcessHandlerReport(True, self.exit_codes.ERROR_KNOWN_UNRECOVERABLE_FAILURE)
 
-        self.ctx.inputs.parameters['ELECTRONS']['diagonalization'] = 'cg'
-        action = 'found diagonalization issues, switching to conjugate gradient diagonalization.'
+        self.ctx.diagonalizations.pop(self.ctx.diagonalizations.index(current_diagonalization))
+        if self.ctx.diagonalizations:
```

Comment:
Actually, I think we could simplify the handler logic just setting `CG` to be by default the last one in ``DiagonalizationType``.